hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -2917,3 +2918,140 @@ func aggregateMachineMessages(msgs []string) string {
+		return ctrl.Result{}, nil
+	}
+	log = log.WithValues("nodePool", nodePoolName)
+
+	nodePool := &hyperv1.NodePool{}
+	if err := r.Client.Get(ctx, supportutil.ParseNamespacedName(nodePoolName), nodePool); err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "error getting nodepool")
+		return ctrl.Result{}, err
+	} else if apierrors.IsNotFound(err) {
+		log.Info("removing secret as nodePool is missing")
+		return ctrl.Result{}, r.Client.Delete(ctx, secret)
+	}
+
+	hcluster, err := GetHostedClusterByName(ctx, r.Client, nodePool.GetNamespace(), nodePool.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	controlPlaneNamespace := manifests.HostedControlPlaneNamespace(hcluster.Namespace, hcluster.Name)
+	expectedCoreConfigResources := expectedCoreConfigResourcesForHostedCluster(hcluster)
+	releaseImage, err := r.getReleaseImage(ctx, hcluster, nodePool.Status.Version, nodePool.Spec.Release.Image)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	config, missingConfigs, err := r.getConfig(ctx, nodePool, expectedCoreConfigResources, controlPlaneNamespace, releaseImage, hcluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if missingConfigs {
+		return ctrl.Result{}, nil
+	}
+	targetVersion := releaseImage.Version()
+
+	pullSecretName, err := r.getPullSecretName(ctx, hcluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	globalConfig, err := globalConfigString(hcluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	targetPayloadConfigHash := payloadConfigHash(config, targetVersion, pullSecretName, globalConfig)
+
+	// synchronously deleting the ignition token is unsafe; we need to clean up tokens by annotating them to expire
+	synchronousCleanup := func(ctx context.Context, c client.Client, secret *corev1.Secret) error {
+		return c.Delete(ctx, secret)
+	}
+	type nodePoolSecret struct {
+		expectedName   string
+		matchingPrefix string
+		cleanup        func(context.Context, client.Client, *corev1.Secret) error
+	}
+	valid := false
+	options := []nodePoolSecret{
+		{
+			expectedName:   TokenSecret(controlPlaneNamespace, nodePool.Name, targetPayloadConfigHash).Name,
+			matchingPrefix: tokenSecretPrefix,
+			cleanup:        setExpirationTimestampOnToken,
+		},
+		{
+			expectedName:   IgnitionUserDataSecret(controlPlaneNamespace, nodePool.GetName(), targetPayloadConfigHash).Name,
+			matchingPrefix: ignitionUserDataPrefix,
+			cleanup:        synchronousCleanup,
+		},
+	}
+	cleanup := synchronousCleanup
+	var names []string
+	for _, option := range options {
+		names = append(names, option.expectedName)
+		if secret.Name == option.expectedName {
+			valid = true
+		}
+		if strings.HasPrefix(secret.Name, option.matchingPrefix) {
+			cleanup = option.cleanup
+		}
+	}
+
+	if valid {
+		return ctrl.Result{}, nil
+	}
+
+	log.WithValues("options", names, "valid", valid).Info("removing secret as it does not match the expected set of names")
+	return ctrl.Result{}, cleanup(ctx, r.Client, secret)
```

Comment:
could this controller ever result in a new valid secret triggering reconciliation, then any of the values used to compute the expectedName coming from a stale cache, and so resulting in pruning a legit secret?

Reply:
The two reconcilers share a manager and cache, so I think it should be possible for one client of that cache to read something more stale than the other.